packages/drivers/odsp-driver/src/fetchWithEpochValidation.ts

```diff
@@ -0,0 +1,125 @@
+        try {
+            const response = await fetchAndParseAsJSONHelper<T>(url, fetchOptions);
+            this.validateEpochFromResponse(response.headers.get("x-fluid-epoch"));
+            return response;
+        } catch (error) {
+            this.checkForEpochError(error);
+            throw error;
+        }
+    }
+
+    public async fetchResponse(
+        url: string,
+        fetchOptions: {[index: string]: any},
+        addInBody: boolean = false,
+    ): Promise<Response> {
+        // Add epoch either in header or in body.
+        this.addEpochInRequest(fetchOptions, addInBody);
+        try {
+            const response = await fetchHelper(url, fetchOptions);
+            this.validateEpochFromResponse(response.headers.get("x-fluid-epoch"));
+            return response;
+        } catch (error) {
+            this.checkForEpochError(error);
+            throw error;
+        }
+    }
+
+    private addEpochInRequest(fetchOptions: {[index: string]: any}, addInBody: boolean) {
+        if (this.fluidEpoch !== undefined) {
+            if (addInBody) {
+                // We use multi part form request for post body where we want to use this.
+                // So extract the form boundary to mark the end of form.
+                let body: string = fetchOptions.body;
+                const formBoundary = body.split("\r\n")[0].substring(2);
+                body += `\r\nepoch=${this.fluidEpoch}\r\n`;
+                body += `\r\n--${formBoundary}--`;
+                fetchOptions.body = body;
+                return;
+            }
+            // Else add in headers.
+            fetchOptions.headers = {
+                ...fetchOptions.headers,
+                "x-fluid-epoch": this.fluidEpoch,
+            };
+        }
+    }
+
+    private validateEpochFromResponse(epochFromResponse: string | undefined | null) {
+        // If epoch is undefined, then don't compare it because initially for createNew or TreesLatest
+        // initializes this value. Sometimes response does not contain epoch as it is still in
+        // implementation phase at server side. In that case also, don't compare it with our epoch value.
+        assert(this.fluidEpoch === undefined
+            || epochFromResponse === undefined || epochFromResponse === null
+            || (this.fluidEpoch === epochFromResponse), "Fluid Epoch should match");
+        if (epochFromResponse !== undefined && epochFromResponse !== null) {
+            this._fluidEpoch = epochFromResponse;
+        }
+    }
+
+    private checkForEpochError(error) {
+        if (error.errorType === OdspErrorType.epochVersionMismatch && this.logger !== undefined) {
+            this.logger.sendErrorEvent({ eventName: "EpochVersionMismatch" }, error);
+            // If the epoch mismatches, then clear all entries for such document from cache.
+            assert(this._hashedDocumentId, "DocId should be set to clear the cached entries!!");
```

Comment:
This is a bit too late check, I.e. this validation should happen before we fail epoch test, to improve chances of hitting a bug and fixing it before changes go into production - very very would test actual epoch changes.
Can we test it unconditionally at the beginning of this function?

Reply:
Replaced.